examples/Consumer/Program.cs

```diff
@@ -61,19 +61,16 @@ public static void Run_Consume(string brokerList, List<string> topics, Cancellat
                 // Note: All handlers are called on the main .Consume thread.
                 .SetErrorHandler((_, e) => Console.WriteLine($"Error: {e.Reason}"))
                 .SetStatisticsHandler((_, json) => Console.WriteLine($"Statistics: {json}"))
-                .SetRebalanceHandler((_, e) =>
-                {
-                    if (e.IsAssignment)
-                    {
-                        Console.WriteLine($"Assigned partitions: [{string.Join(", ", e.Partitions)}]");
-                        // possibly override the default partition assignment behavior:
-                        // consumer.Assign(...) 
-                    }
-                    else
-                    {
-                        Console.WriteLine($"Revoked partitions: [{string.Join(", ", e.Partitions)}]");
-                        // consumer.Unassign()
-                    }
+                .SetPartitionAssignmentHandler((c, partitions) =>
+                {
+                    Console.WriteLine($"Partition assignment: [{string.Join(", ", partitions)}]");
+                    // possibly override the default partition assignment behavior:
+                    // c.Assign(...) 
+                })
+                .SetPartitionAssignmentRevokedHandler((c, partitions) =>
```

Comment:
I'm interested in @ybyzek 's take on those names too. certainly agree I'd like better ones (but racking my brain yielded nothing, for this structure).

I very much like the word rebalance, but I went sour on it with the current structure this PR is replacing ( for reasons outlined https://github.com/confluentinc/confluent-kafka-dotnet/pull/789#issue-254382118 ). I think it works as in this suggestion https://github.com/confluentinc/confluent-kafka-dotnet/pull/789#issuecomment-465303319

Another option is `SetRebalanceHandlers`, taking two arguments - the assignment and assignmentRevoked delegates.

Another option is `SetRebalanceHandler` taking two arguments. The first an enum, either `RebalanceType.Assignment` or `RebalanceType.AssignmentRevoked`, the second the handler. If you want to set both you call it twice.

I trialed all the above and arrived at the current PR as my favorite, but i could do that exercise again to see if i still arrive at the same conclusion.

note that in all the above, I believe `Cosumer.Assignment` should be changed to `Cosumer.AssignedPartitions`.

Reply:
@edenhill - thought through all the alternatives again, and opened #805.

Reply:
At first glance it does read a bit strangely.  I don't have much context for the APIs but can research and get back to you.  Meanwhile, for your consideration, you could pull in some expert advice from Kafka core?

Reply:
thanks @ybyzek - yep, core are in the loop. I'm interested also in your first reaction to the alternative in https://github.com/confluentinc/confluent-kafka-dotnet/pull/805

Reply:
* `SetPartitionsAssignedHandler` (rather than `SetPartitionsAssignmentHandler`) seems to better fit `SetPartitionsRevokedHandler`
* Regarding the optimization of combining into one call but pass in different parameters (https://github.com/confluentinc/confluent-kafka-dotnet/pull/805): it depends how often they'd be used together.  If it's a common pattern that users would call one but not the other, then it may be more beneficial to keep them separate and call them with their own explicit methods.
* How important is it to have parity with Java counterparts?   For broader adoption, consistency does help to reduce friction in adoption of a new client, but I gather @mhowlett disagrees

> I think the suggestion that we adopt Java terminology because that will make things easier for users is backwards - I think we should use different terminology to alert users that things are a bit different with this client.

Reply:
@ybyzek - the behavior of librdkakfa is different from the java client, and this necessarily transfers through to this client. i'd prefer to use the same terminology as the java client, but doing so when the functionality is subtly different doesn't bring consistency, it obfuscates.

Reply:
(is my point with that comment). I think your feedback reinforces this - it will help even people with a lot of familiarity with Kafka in preventing confusion between this API and Java API.

Reply:
The tradeoff with the 0.11.6 naming: the names sound good and familiar, but it's not precise and I believe quite confusing.
The tradeoff with this PR: it resolves the ambiguity, but only if you look at it the right way (it's not super clear assignment and assign/assigned are separate things) and the names are more awkward.
The tradeoff with the #805: it avoids use of a word stemming from assign to remove ambiguity in the handler names, but it's preferable to not have to specify the two handlers in the same method.

Maybe if assign/revoke are qualified, it makes it ok:

`SetRebalanceAssignmentHandler`, `SetRebalanceRevokationHandler`
`SetRebalanceAssignHandler`, `SetRebalanceRevokeHandler`
`SetCoordinatorAssignmentHandler`, `SetCoordinatorRevocationHandler`
`SetGroupCoordinatorAssignmentHandler`, `SetGroupCoordinatorRevocationHandler`

not really liking any of that.

Reply:
Another idea: it would be possible to make this work like the Java consumer - you could intercept calls to `Seek` and work out if they happened in a partition assignment handler. if they did, you could call assign behind the scenes instead of seek. you wouldn't need to give up on librdkafka's capability to assign to more partitions than the group assignment - the additional partitions could be specified during configuration (not during the callback handler which is weird imo), and the set of partitions passed into the assignment handler would be the group assignment + the configured extra ones. The names `SetPartitionsAssignedHandler`, `SetPartitionsRevokedHandler` and `Assignment` would then all be good.

the new idea here is intercepting calls to seek to avoid calling assign in the partitions assigned handler (which was a show stopper for me for this idea previously).

could easily be missing something.

Reply:
Continuing the previous comment: in the abstraction exposed to the user, the partitions have already been 'assigned' when the handler is called (in both senses, making the the name `SetPartitionsAssignedHandler` very good), but they are paused - no fetching starts until after the handler has finished.

It's not ideal that the C# binding warp the abstraction provided by librdkafka into something different. However, countering that is making the behaviour match the Java client, which is arguably a bigger win. My main concern in doing this would be if the abstraction is leaky in some important way I haven't realised.

In the 1.0 scope, i'd propose not providing the capability for additional partitions to be added to the group coordinator assignment. The API allows for this capability to be added though in the future.

@edenhill - v. keen for your point of view on this.